pithos/plugins/notify.py

```diff
@@ -52,16 +55,24 @@ def send_notification(self, window, *ignore):
             self._app.withdraw_notification(self._app_id)
         else:
             song = window.current_song
+            album = '\n' + song.album if song.album else ''
             # This matches GNOME-Shell's format
             notification = Gio.Notification.new(song.artist)
             # GNOME focuses the application by default, we want to match that behavior elsewhere such as on KDE.
             notification.set_default_action('app.activate')
-            notification.set_body(song.title)
+            notification.set_body(song.title + album)
 
+            # Use BytesIcon for GNOME and Flatpak and ThemedIcon everywhere else, otherwise notifications do not work
```

Comment:
I still really hate this. Its very wrong that this works this way.

Let's do:

```python
notify_server_name = ... 

if is_flatpak():
  # Use Gio.BytesIcon. Anything else from a sandbox is nonsense.
else:
  notify_server_name = ...# Get the result of `org.freedesktop.Notifications.GetServerInformation()` (This should be async, so that adds some complications)
  if notify_server_name in ('whatever', ): # A list of daemons where Gio.FileIcon does not work?
    # FIXME: This is a weird hack where we shove a file URI in a themed icon and some daemons accept this and use it as they do not support FileIcon.
    # Use Gio.ThemedIcon
  else:
    # Use Gio.FileIcon
```

Reply:
Also I'm sorry if I keep repeating myself. But at the end of the day it is *not* Pithos' bug that `Gio.BytesIcon` doesn't work somewhere. That is a real bug outside of Pithos and should be fixed there.

`Gio.ThemedIcon` working at all is truly a hack. One I can accept as long as its reported upstream where the real bug is, limited in scope only to where it is known to work, and documented for what it is.

Reply:
I'm not 100% sure where the bug should be filed upstream or what details would need to be provided. I understand and agree that a longer-term solution is to address the upstream bug rather than work around it but it feels outside of the scope of addressing the problem in the present. I can open an issue where needed and provide what I can but it is unlikely I would be able to own it which gives me further hesitation. (I'm not a GNOME user primarily after all.)

And the scope of working and not working is inverted to my understanding. `BytesIcon` doesn't "work" in GNOME, it just doesn't prevent the notification from being displayed. On Cinnamon and Enlightenment (unsure of the scope outside of these two), it causes notifications not to be shown **at all** when Pithos is installed via Flatpak, negatively impacting the end user experience as evidenced by the open Issues.

Breaking this further down into two issues:

1. Address the bug with notifications in the current application as it is affecting end users.
2. Implement the fix in line with the expectations of the available libraries.

The fix for 1 is to use `FileIcon` for the version of Pithos packaged outside of Flatpak. When running as a Flatpak, if GNOME is the desktop, use `BytesIcon` and in all other Flatpak conditions, use `ThemedIcon` for now, as the scope is everything not GNOME that is impacted. (I also realize this was not written into the conditionals well in the latest commit but I am working on the requested changes now as well as correcting the logic to match.) This fixes the issue immediately for impacted end users.

When the libraries are patched upstream and known to be working as designed, we can address 2 and should revisit the code and re-implement it as it is intended to function. That change would and should be both transparent to the end user and free from regression.

Can you agree on this strategy and path forward?

Reply:
@TingPing Can you share your thoughts on the latest push and the previous comments above? The alternative approach you've suggested is the inverse strategy to solve this problem. It might be better to test or solicit feedback from more desktop environments before concluding that GNOME is the norm rather than the exception, maybe via a pinned Issue for users to submit feedback. The idea of a tuple (or list or array or set) that would need to be maintained until some upstream bug is resolved is a significantly more tiresome and errorful approach. Between KDE Plasma, MATE, XFCE, LXDE, and many others, this alternative approach potentially leaves them in a broken state which is not the preferred strategy.

If it would be easier and your schedule is amenable to it, we can set a time to discuss this over something like Jitsi. Or we can continue with this thread as is. The main motives being to ensure that we are on the same page and that we can reach an unbiased conclusion in a timely fashion.

Reply:
At this point I'll accept whatever works in the most places. It is a frustrating problem.